### ISSUE TITLE
Bitstring Status List v1.0 2024-01-21 > 2024-02-13

### ISSUE THREAD

**Issue by msporny** [w3cping/privacy-request#127]
- Name of spec to be reviewed: 
  - Bitstring Status List v1.0
- URL of spec: 
  - https://w3c.github.io/vc-bitstring-status-list/
- What and when is your next expected transition?
  - Candidate Recommendation, February 2024
- What has changed since any previous review?
  - This is the first review performed by your group
- Please point to the results of your own self-review (see https://w3ctag.github.io/security-questionnaire/, https://w3c.github.io/fingerprinting-guidance/, https://tools.ietf.org/html/rfc6973)
  - https://github.com/w3c/vc-bitstring-status-list/issues/77
- Where and how to file issues arising?
  - https://github.com/w3c/vc-bitstring-status-list/issues
- Pointer to any explainer for the spec?
  -  https://github.com/w3c/vc-bitstring-status-list/blob/main/explainer.md

### Other comments:

Our apologies, we thought we had submitted this review when we [submitted the TAG review request](https://github.com/w3ctag/design-reviews/issues/874), and when [we did the security and privacy self-review](https://github.com/w3c/vc-bitstring-status-list/issues/77), but we failed to do so. 

We were hoping to enter CR next month, but are gated by this review (the WG believes it is done enough to enter CR).

**Comment by msporny** [comment]
@kdenhartog wrote:
> I'll assist on this one with @npdoty

Great, thank you @kdenhartog ! :)

**Comment by npdoty** [comment]
[PING discussed this on our February 1 call](https://www.w3.org/Privacy/IG/summaries/PING-minutes-20240201#bitstring-status-list-from-kyle-and-nick).

Summary of issues:
(@kdenhartog, @jyasskin and others can review to see if I've got the right content here. VCWG, I can make separate issues with these once someone has reviewed my notes, but if you want to look at them ASAP, feel free and just recognize that this is one reviewer's imperfect notes.)

The status list index appears to be a cross-origin persistent unique identifier. The same status list index provided on a credential used on two different websites could let those websites link the user between them. This will be abused as a very effective cross-context tracking vector.

An issuer could provide unique status list URLs per credential, in order to effectively track when a holder's credential is used at each verifier. (This threat is already noted in the privacy considerations.) We should consider mitigations for this threat, including consistency checks. This [draft on key consistency and discovery](https://datatracker.ietf.org/doc/draft-ietf-privacypass-key-consistency/) may be a useful starting point.

Caching and CDNs are suggested as potential mitigations for hiding when a holder visited a verifier. Caching recommendations to verifiers could use normative guidance. It's not clear that CDNs provide a meaningful privacy protection; rather, anonymizing proxies could provide a more explicit design for that purpose. [Oblivious HTTP](https://datatracker.ietf.org/doc/rfc9458/) might be a better fit.

Checking for revocation or status updates in general also reveal information about the holder to the verifier, potentially indefinitely afterwards. This is a significant risk, especially for open-ended status messages, or where status could reveal something sensitive (revocation of a privilege, 

Larger issues: 

Revocation status is often not important to the use case for credentials. For example, my driver's license may have expired or my driving privileges may have been revoked, but my age won't have changed in either case. It seems guidance is necessary for when status list information should be included at all in the response to a credential request, and when it isn't appropriate. Or when it is appropriate, the spec needs to highlight the privacy issues in doing so, and verifiers and holder software will need to communicate that to the user.

There are alternative designs for revocation that may have different privacy properties, and might be preferable. Considering these would be a longer work item, but, for example, a stapled validity proof would be a way for a holder to prove that a credential is recently valid. Short-lived credentials could provide assurance to the verifier when necessary -- requiring the holder to more frequently interact with the issuer, but under the holder's control and without involving the particular verifier. Cryptographic accumulators could be used to efficiently communicate revocation/non-revocation status without requiring the verifier to contact the issuer.

https://ieeexplore.ieee.org/document/9376112
https://medium.com/@alexeysamoshkin/how-ssl-certificate-revocation-is-broken-in-practice-af3b63b9cb3

Smaller issues:

Could dummy entries in the status list help to protect against an attacker who just regularly scans the list to try to reidentify users or their status?

The spec requires that the status list index SHOULD be randomized. I can see that there may be alternatives, but there could be a MUST requirement for the unpredictability/intelligence-free nature of the index: it must not be something that a recipient could infer, or where the index reveals something about the credential (like its recency, etc.).

**Comment by martinthomson** [comment]
In addition to the comments from @npdoty, I will note that this design depends on indices of issued credentials being sequential.  That has secondary concerns from the perspective of privacy.  If I can observe when someone is issued a credential, I can use this information to determine whether their credential is revoked.  Though this concern is just a variant of the "persistent unique identifier" concern above, the point is that the design depends on the value for other people being highly predictable.

On the design front, did the working group consider designs like [CRLite](https://crlite.jameslarisch.com/)?  If the universe of possible values is known, then a probabilistic data structure can be used to compress even more.  And - though I haven't verified this - the privacy issues that are expressed above might not apply in that context.

(Also, from a pure information theory perspective, zlib might be as effective as a far simpler run-length encoding, but I'm not sure that it would be in practice if you have decent coding for the lengths.)

**Comment by npdoty** [comment]
I've opened issues on the `vc-bitstring-status-list` repo:
https://github.com/w3c/vc-bitstring-status-list/issues/created_by/npdoty
Apologies for the delay.

**Comment by msporny** [comment]
> I've opened issues on the `vc-bitstring-status-list` repo: https://github.com/w3c/vc-bitstring-status-list/issues/created_by/npdoty Apologies for the delay.

Thank you! We will start processing these next week. We were hoping to move Bitstring Status List to CR by mid-to-end-April 2024; please try to help us achieve that by engaging as we have discussions and raise PRs. We are treating all issues PING raised as needing to be addressed by the WG before we enter CR.

**Comment by msporny** [comment]
All issues raised by PING have been addressed via PRs that were merged into the specification: 

https://github.com/w3c/vc-bitstring-status-list/issues?q=is%3Aissue+is%3Aclosed+label%3Aprivacy-needs-resolution

We will be requesting transition to CR next week, please be advised that W3M will ask for status for this issues shortly if it isn't closed. If we missed anything, please let us know so that we can address any further comments.
